dart/lib/src/builder.dart

```diff
@@ -0,0 +1,623 @@
+import 'dart:convert';
+import 'dart:typed_data';
+
+import 'types.dart';
+
+/// The main builder class for creation of a FlexBuffer
+class Builder {
+  ByteData _buffer;
+  List<_StackValue> _stack;
+  List<_StackPointer> _stackPointers;
+  int _offset;
+  bool _finished;
+  Map<String, _StackValue> _stringCache;
+  Map<String, _StackValue> _keyCache;
+  Map<_KeysHash, _StackValue> _keyVectorCache;
+  Map<int, _StackValue> _indirectIntCache;
+  Map<double, _StackValue> _indirectDoubleCache;
+
+  /// Instantiate the builder if you intent to gradually build up the buffer by calling
+  /// add... methods and calling finish() to receive the the resulting byte array.
+  /// The default size of internal buffer is set to 2048. Provide a different value in order to avoid buffer copies.
+  Builder([int size = 2048]) {
+    _buffer = ByteData(size);
+    _stack = [];
+    _stackPointers = [];
+    _offset = 0;
+    _finished = false;
+    _stringCache = {};
+    _keyCache = {};
+    _keyVectorCache = {};
+    _indirectIntCache = {};
+    _indirectDoubleCache = {};
+  }
+
+  /// Use this method in order to turn an object into a FlexBuffer directly.
+  /// Please use the manual instantiation of the [Builder] and gradual addition of values, if performance is more important than convenience.
+  static ByteBuffer buildFromObject(dynamic value) {
+    var flx = Builder();
+    flx._add(value);
+    var buffer = flx.finish();
+    var bd = ByteData(buffer.lengthInBytes);
+    bd.buffer.asUint8List().setAll(0, buffer);
+    return bd.buffer;
+  }
+
+  void _add(dynamic value) {
+    if (value == null) {
+      addNull();
+    } else if (value is bool) {
+      addBool(value);
```

Comment:
explicitly cast - this will be important for NNBD applications coming up.

Reply:
(i.e. `value as bool`)

Reply:
Could you please sketch an example? 
What I can think of is something like:
```
    final b = value as bool;
    if (b != null) {
      addBool(b);
      return;
    }
    final i = value as int;
    if (i != null) {
      addInt(i);
      return;
    }
    ...
```
IMHO, current implementation is easier on the eye and also correct in regards to NNBD as the first `if` statement checks for `null` explicitly.

Reply:
I may be wrong about this - the compiler might have enough information here to just figure things out for us. Let's leave it as is, and whenever we do null safety for real we can re-evaluate it.